### PR TITLE
Add task tracking option for heap usage monitoring

### PR DESCRIPTION
**Pull request by slcasner**
Add back a feature that was available in the old heap implementation
in release/v2.1 and earlier: keep track of which task allocates each
block from the heap.  The task handle is conditionally added as
another word in the heap poisoning header under this configuration
option CONFIG_HEAP_TASK_TRACKING.

To allow custom monitoring and debugging code to be added, move data
structure declarations from multi_heap.c and multi_heap_poisoning.c
into multi_heap_internal.h and a new file multi_heap_poisoning.h,
respectively.

As a result, the struct multi_heap_info with typename heap_t conflicts
with heap_t declared in heap_private.h for code that needs both
headers.  Also, it might be confusing to have this struct named
multi_heap_info when there is a different struct in multi_heap.h with
the typename multi_heap_info_t.  Therefore struct multi_heap_info was
renamed multi_heap_metadata with typename heap_metadata_t.

### CODE REVIEW

File: components/heap/multi_heap_poisoning.c [hunk 1]
```diff
@@ -67,6 +59,9 @@ static uint8_t *poison_allocated_region(poison_head_t *head, size_t alloc_size)
     poison_tail_t *tail = (poison_tail_t *)(data + alloc_size);
     head->alloc_size = alloc_size;
     head->head_canary = HEAD_CANARY_PATTERN;
+#ifdef CONFIG_HEAP_TASK_TRACKING
+    head->task = xTaskGetCurrentTaskHandle();
```
Comment (projectgus) [inline components/heap/multi_heap_poisoning.c]: One of the (probably unwritten, sorry) design decisions for multi_heap is that it doesn't contain any platform-specific code. This allows us to run, for example, the host-side tests in test_multi_heap_host. Eventually we may split multi_heap out to a standalone project outside IDF.

Would you be able to please move the platform-specific parts of this (config macro and FreeRTOS API call) to multi_heap_config.h and multi_heap_platform.h, respectively? Also, please check the host-side tests still compile & pass.

File: components/heap/multi_heap_internal.h [hunk 2]
```diff
@@ -45,3 +45,41 @@ void multi_heap_internal_poison_fill_region(void *start, size_t size, bool is_fr
 void multi_heap_internal_lock(multi_heap_handle_t heap);
 
 void multi_heap_internal_unlock(multi_heap_handle_t heap);
+
+struct heap_block;
```
Comment (projectgus) [inline components/heap/multi_heap_internal.h]: Can you please give some examples the internal debugging code you've written to use this? Even though this header as marked as "internal" and not added to the public include path list, I'm a little nervous about exposing the heap implementation internals.

Maybe there are ways we can expand the existing multi_heap APIs to expose useful debug operations in a more practical way? Perhaps some sort of get_alloc_info(ptr, &info) which will return info about the heap block containing 'ptr'? Ideas (based on your own debug usage) much appreciated. :)

**Comment by slcasner** [comment]
@prjectgus, thanks for the prompt comments.  I will think about the changes you suggest and respond in more detail.  I will also be committing the code for my use case.  Meanwhile, here is an example of how this information is displayed in our app.  The top section shows the memory in use and +/- change for all tasks that have added or freed memory since the previous instance of the command.  The lower section shows all the individual blocks added or freed if a task name is included on the command (like tiT in this example).

    OVMS > module memory tiT
    ============================
    Free 8-bit 74380/147632, 32-bit 21988/48772, blocks dumped = 27
    task=Housekeeping    total=  26630  33762  26020 change=    -96  -4569     +0
    task=tiT             total=    108   1436      0 change=    +76  +1184     +0
    task=wifi            total=      0   1772      0 change=     +0   +312     +0
    task=eventTask       total=      8  13151      0 change=     +8 +13151     +0
    task=NetManTask      total=      0   2262      0 change=     +0  +2262     +0
    ============================
    + t=tiT s=  16 a=0x3ffcc3fc  3FFE369C 00000190 401B5E68 00000000 BAAD5678 3FFCC434 ABBA1234 3FFCD394
    + t=tiT s=  16 a=0x3ffcc420  3FFE5E74 00000064 401B5E28 00000000 BAAD5678 3FFCC454 ABBA1234 3FFCD394
    + t=tiT s=  12 a=0x3ffcc444  3FFE5FF0 3FFE578C 2E696601 BAAD5678 3FFCC461 3FFCC6F0 0007BAAD 3FFCC47C
    + t=tiT s=  16 a=0x3ffcc48c  3FFCC4B0 00000000 401B5E14 00000000 BAAD5678 3FFCC4C4 ABBA1234 3FFCD394
    + t=tiT s=  16 a=0x3ffcc4b0  3FFCC4D4 00000000 401B5E00 00000000 BAAD5678 3FFCC4E8 ABBA1234 3FFCD394
    + t=tiT s=  16 a=0x3ffcc4d4  3FFCC4F8 00000000 401B5DEC 00000000 BAAD5678 3FFCC50C ABBA1234 3FFCD394
    + t=tiT s=  16 a=0x3ffcc4f8  00000000 00000000 401B5E3C 00000000 BAAD5678 3FFCC52C ABBA1234 3FFD7A10
    + t=tiT s=  96 a=0x3ffe35f8  1197AB56 00000000 01000A01 00000000 00000000 00000044 0000A8C0 00012750
    + t=tiT s=  16 a=0x3ffe369c  3FFCC48C 00000000 401B5E50 00000000 BAAD5678 3FFE371C ABBA1234 3FFBDFAC
    + t=tiT s=  20 a=0x3ffe3844  3FFE5FC8 3FFEAC18 FB0000E0 00000200 00000001 BAAD5678 3FFE3871 3FFE3F48
    + t=tiT s=  88 a=0x3ffe3f9c  00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
    + t=tiT s=  88 a=0x3ffe578c  00000000 00000000 3FFE578C 00000000 00000000 3FFE57A4 FFFFFFFF 3FFE57A4
    + t=tiT s=  88 a=0x3ffe5960  00000000 00000000 3FFE5960 00000000 00000000 3FFE5978 FFFFFFFF 3FFE5978
    + t=tiT s=  68 a=0x3ffe5ac0  00000000 312B203D 36393031 20202020 20302B00 20202020 740A302B 3D6B7361
    + t=tiT s=  88 a=0x3ffe5b60  00000000 00000000 3FFE5B60 00000000 00000000 3FFE5B78 FFFFFFFF 3FFE5B78
    + t=tiT s=  32 a=0x3ffe5bcc  00000000 3FFEAC18 000002FF 00000000 01000000 000000FF 00000200 3FFB7601
    + t=tiT s=  68 a=0x3ffe5d98  00000000 000000A0 00000000 00000000 00000000 00000000 00000000 00000000
    + t=tiT s= 112 a=0x3ffe5df0  3FFE5E48 3FFE5E60 3FFE5E48 3FFE5E5C 00000000 3FFE5E08 FFFFFFFF 3FFE5E08
    + t=tiT s=  16 a=0x3ffe5e74  3FFCC3FC 00000000 401B5DD8 00000000 BAAD5678 3FFE5EB0 ABBA1234 3FFD7A10
    + t=tiT s=  20 a=0x3ffe5fc8  00000000 3FFEAC18 010000E0 00000200 66206901 BAAD5678 3FFE6064 ABBA1234
    + t=tiT s= 112 a=0x3ffe5ff0  3FFE6048 3FFE6060 3FFE6048 3FFE605C 00000000 3FFE6008 FFFFFFFF 3FFE6008
    + t=tiT s=  12 a=0x3ffe6074  3FFE616C 3FFE5960 00006D01 BAAD5678 3FFE60F0 ABBA1234 3FFEAD14 00000058
    + t=tiT s= 112 a=0x3ffe616c  3FFE61C4 3FFE61DC 3FFE61C4 3FFE61D8 00000000 3FFE6184 FFFFFFFF 3FFE6184
    + t=tiT s=  68 a=0x3ffe6360  00000000 496D3233 37322820 20293933 73626500 65767265 4C203A72 636E7561
    + t=tiT s=  12 a=0x3ffe7cd4  3FFE5DF0 3FFE5B60 00000001 BAAD5678 3FFE7CED 3FFE96D8 3FFE7D38 ABBA1234
    + t=tiT s=  88 a=0x3ffea5c8  5401A8C0 00000000 00000000 00000000 00000000 FB0000E0 00000000 00000000
    + t=tiT s= 232 a=0x3ffeac18  3FFB8E88 5401A8C0 00000000 00000000 00000000 00000000 00FFFFFF 00000000
    ============================

**Comment by slcasner** [comment]
Replaced by PR 1487 on master that incorporates the feedback from here.

**Pull request closed without merging**